File: ports/lua/test/form_and_bip39_test.lua
local h = require("test_helper")
local locale = require("unicode_lua.security.form.locale_case_inversion")
local nfc = require("unicode_lua.security.form.nfc_idempotence_witness")
local bomb = require("unicode_lua.security.form.normalization_bomb")
local bip39 = require("unicode_lua.security.crypto.bip39_canonical")

local function assert_nil(v, label)
  if v ~= nil then
    error(label .. " expected nil")
  end
end

assert_nil(locale.detect({}).sub, "locale empty")
assert_nil(locale.detect({ 0x48, 0x65, 0x6C, 0x6C, 0x6F }).sub, "locale ascii")
h.assert_equal(locale.detect({ 0x0049 }).sub, "TurkishCaseDivergence", "locale I")
h.assert_equal(locale.detect({ 0x0049 }).positions, { 0 }, "locale I pos")
h.assert_equal(locale.detect({ 0x0130 }).sub, "TurkishCaseDivergence", "locale dotted I")
h.assert_equal(locale.detect({ 0x0049, 0x0300 }).sub, "TurkishCaseDivergence", "locale priority")
h.assert_equal(locale.detect({ 0x004A, 0x0300 }).sub, "LithuanianCaseDivergence", "locale lt")

assert_nil(nfc.detect({}).sub, "nfc empty")
assert_nil(nfc.detect({ 0x48, 0x65, 0x6C, 0x6C, 0x6F }).sub, "nfc ascii")
assert_nil(nfc.detect({ 0x00E9 }).sub, "nfc composed")
h.assert_equal(nfc.detect({ 0x0065, 0x0301 }).sub, "NonNfcForm", "nfc decomposed")
h.assert_equal(nfc.detect({ 0x0065, 0x0301 }).positions, { 0 }, "nfc pos")
h.assert_equal(nfc.detect({ 0xFB01 }).sub, "NonNfkcCompatForm", "nfkc ligature")

assert_nil(bomb.detect({}).sub, "bomb empty")
assert_nil(bomb.detect({ 0x48, 0x65, 0x6C, 0x6C, 0x6F }).sub, "bomb ascii")
assert_nil(bomb.detect({ 0xD55C }).sub, "bomb korean")
assert_nil(bomb.detect({ 0x2460 }).sub, "bomb circled")
h.assert_equal(bomb.detect({ 0xFDFA }).sub, "SingleCpBlowup", "bomb blowup")
h.assert_equal(bomb.detect({ 0xFDFA }).positions, { 0 }, "bomb blowup pos")
h.assert_equal(bomb.detect({ 0xFDFB }).sub, "NfkdHighExpansion", "bomb nfkd")
h.assert_equal(bomb.detect({ 0x1F82 }).sub, "NfdHighExpansion", "bomb nfd")

local abandon = { 0x61, 0x62, 0x61, 0x6E, 0x64, 0x6F, 0x6E }
local about = { 0x61, 0x62, 0x6F, 0x75, 0x74 }
h.assert_equal(bip39.bip39_canonical({ 0x61, 0x20, 0x20, 0x62 }), { 0x61, 0x20, 0x62 }, "bip39 collapse")
h.assert_equal(bip39.bip39_canonical({ 0x41 }), { 0x61 }, "bip39 lower")
h.assert_equal(bip39.detect({ table.unpack(abandon), 0x20 }).sub, "TrailingWhitespace", "bip39 trailing")
h.assert_equal(bip39.detect({ 0x41, 0x62, 0x61, 0x6E, 0x64, 0x6F, 0x6E }).sub, "MixedCase", "bip39 case")
local double_ws = { table.unpack(abandon) }
double_ws[#double_ws + 1] = 0x20
double_ws[#double_ws + 1] = 0x20
for _, cp in ipairs(about) do double_ws[#double_ws + 1] = cp end
h.assert_equal(bip39.detect(double_ws).sub, "WhitespaceAnomaly", "bip39 ws")
h.assert_equal(bip39.detect({ 0xFB00 }).sub, "NonNFKD", "bip39 nfkd")
h.assert_equal(bip39.detect({ 0x71, 0x7A, 0x71, 0x7A }).sub, "WordlistMismatch", "bip39 word")
local mnemonic = {}
for _ = 1, 11 do
  for _, cp in ipairs(abandon) do mnemonic[#mnemonic + 1] = cp end
  mnemonic[#mnemonic + 1] = 0x20
end
for _, cp in ipairs(about) do mnemonic[#mnemonic + 1] = cp end
local verdict = bip39.detect(mnemonic)
assert_nil(verdict.sub, "bip39 clear")
h.assert_equal(verdict.language, "english", "bip39 lang")
h.assert_equal(verdict.word_count, 12, "bip39 count")

File: ports/php/test/form_and_bip39_test.php
<?php

declare(strict_types=1);

require_once __DIR__ . '/test_helper.php';

use UnicodePhp\Security\Crypto\Bip39Canonical;
use UnicodePhp\Security\Form\LocaleCaseInversion;
use UnicodePhp\Security\Form\NfcIdempotenceWitness;
use UnicodePhp\Security\Form\NormalizationBomb;

assert_same_value(null, LocaleCaseInversion::detect([])->sub, 'locale empty');
assert_same_value(null, LocaleCaseInversion::detect([0x48, 0x65, 0x6C, 0x6C, 0x6F])->sub, 'locale ascii');
assert_same_value('TurkishCaseDivergence', LocaleCaseInversion::detect([0x0049])->sub, 'locale I');
assert_same_value([0], LocaleCaseInversion::detect([0x0049])->positions, 'locale I pos');
assert_same_value('TurkishCaseDivergence', LocaleCaseInversion::detect([0x0130])->sub, 'locale dotted I');
assert_same_value('TurkishCaseDivergence', LocaleCaseInversion::detect([0x0049, 0x0300])->sub, 'locale priority');
assert_same_value('LithuanianCaseDivergence', LocaleCaseInversion::detect([0x004A, 0x0300])->sub, 'locale lt');

assert_same_value(null, NfcIdempotenceWitness::detect([])->sub, 'nfc empty');
assert_same_value(null, NfcIdempotenceWitness::detect([0x48, 0x65, 0x6C, 0x6C, 0x6F])->sub, 'nfc ascii');
assert_same_value(null, NfcIdempotenceWitness::detect([0x00E9])->sub, 'nfc composed');
assert_same_value('NonNfcForm', NfcIdempotenceWitness::detect([0x0065, 0x0301])->sub, 'nfc decomposed');
assert_same_value([0], NfcIdempotenceWitness::detect([0x0065, 0x0301])->positions, 'nfc pos');
assert_same_value('NonNfkcCompatForm', NfcIdempotenceWitness::detect([0xFB01])->sub, 'nfkc ligature');

assert_same_value(null, NormalizationBomb::detect([])->sub, 'bomb empty');
assert_same_value(null, NormalizationBomb::detect([0x48, 0x65, 0x6C, 0x6C, 0x6F])->sub, 'bomb ascii');
assert_same_value(null, NormalizationBomb::detect([0xD55C])->sub, 'bomb korean');
assert_same_value(null, NormalizationBomb::detect([0x2460])->sub, 'bomb circled');
assert_same_value('SingleCpBlowup', NormalizationBomb::detect([0xFDFA])->sub, 'bomb blowup');
assert_same_value([0], NormalizationBomb::detect([0xFDFA])->positions, 'bomb blowup pos');
assert_same_value('NfkdHighExpansion', NormalizationBomb::detect([0xFDFB])->sub, 'bomb nfkd');
assert_same_value('NfdHighExpansion', NormalizationBomb::detect([0x1F82])->sub, 'bomb nfd');

$abandon = [0x61, 0x62, 0x61, 0x6E, 0x64, 0x6F, 0x6E];
$about = [0x61, 0x62, 0x6F, 0x75, 0x74];
assert_same_value([0x61, 0x20, 0x62], Bip39Canonical::bip39Canonical([0x61, 0x20, 0x20, 0x62]), 'bip39 collapse');
assert_same_value([0x61], Bip39Canonical::bip39Canonical([0x41]), 'bip39 lower');
assert_same_value('TrailingWhitespace', Bip39Canonical::detect(array_merge($abandon, [0x20]))->sub, 'bip39 trailing');
assert_same_value('MixedCase', Bip39Canonical::detect([0x41, 0x62, 0x61, 0x6E, 0x64, 0x6F, 0x6E])->sub, 'bip39 case');
assert_same_value('WhitespaceAnomaly', Bip39Canonical::detect(array_merge($abandon, [0x20, 0x20], $about))->sub, 'bip39 ws');
assert_same_value('NonNFKD', Bip39Canonical::detect([0xFB00])->sub, 'bip39 nfkd');
assert_same_value('WordlistMismatch', Bip39Canonical::detect([0x71, 0x7A, 0x71, 0x7A])->sub, 'bip39 word');
$mnemonic = [];
for ($i = 0; $i < 11; $i++) {
    foreach ($abandon as $cp) {
        $mnemonic[] = $cp;
    }
    $mnemonic[] = 0x20;
}
foreach ($about as $cp) {
    $mnemonic[] = $cp;
}
$verdict = Bip39Canonical::detect($mnemonic);
assert_same_value(null, $verdict->sub, 'bip39 clear');
assert_same_value('english', $verdict->language, 'bip39 lang');
assert_same_value(12, $verdict->wordCount, 'bip39 count');
